### PR TITLE
SHA-512: impl SHA-512/224, SHA-512/256

### LINKED ISSUE #61: Is there any interest in adding the SHA512/224 and SHA512/256 algorithms?
I would be happy to do the work, I figure the module would end up exporting the following new objects:

- sha512_256
- sha2_512_256
- SHA2_512_256_CTX

and

- sha512_224
- sha2_512_224
- SHA2_512_224_CTX

The algorithms appear to mostly be the SHA512 with a different initial hash value. Does that seem reasonable?

### PR DESCRIPTION
**Pull request by inkydragon**
Close #61

- ref doc: "FIPS 180-4: Secure Hash Standard (SHS)"
    https://csrc.nist.gov/pubs/fips/180-4/upd1/final

- Reference impl: `pycryptodome`
	https://pycryptodome.readthedocs.io/en/latest/src/hash/sha512.html

1. install `pip install pycryptodome`
2. gen ref hash values:

<details>
<summary>python scripts</summary>

```python
from Crypto.Hash import SHA512

def sha512(s):
    "Input bytes: b'abc'"
    h = SHA512.new()
    h.update(s)
    return h.hexdigest()

def sha512_224(s):
    h = SHA512.new(truncate="224")
    h.update(s)
    return h.hexdigest()

def sha512_256(s):
    h = SHA512.new(truncate="256")
    h.update(s)
    return h.hexdigest()

funcs = [sha512, sha512_224, sha512_256]
lorem = "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
datas = ["", "test", lorem, '\0', 'a'*1000_000]

# [ (f, f(bytes(data, encoding='ASCII'))) for data in datas for f in funcs ]

[ print(f(bytes(data, encoding='ASCII'))) for data in datas for f in [sha512] ]
[ print(f(bytes(data, encoding='ASCII'))) for data in datas for f in [sha512_224] ]
[ print(f(bytes(data, encoding='ASCII'))) for data in datas for f in [sha512_256] ]
```
</details>

Output
```py
>>> [ print(f(bytes(data, encoding='ASCII'))) for data in datas for f in [sha512] ]
cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e
ee26b0dd4af7e749aa1a8ee3c10ae9923f618980772e473f8819a5d4940e0db27ac185f8a0e1d5f84f88bc887fd67b143732c304cc5fa9ad8e6f57f50028a8ff
f41d92bc9fc1157a0d1387e67f3d0893b70f7039d3d46d8115b5079d45ad601159398c79c281681e2da09bf7d9f8c23b41d1a0a3c5b528a7f2735933a4353194
b8244d028981d693af7b456af8efa4cad63d282e19ff14942c246e50d9351d22704a802a71c3580b6370de4ceb293c324a8423342557d4e5c38438f0e36910ee
e718483d0ce769644e2e42c7bc15b4638e1f98b13b2044285632a803afa973ebde0ff244877ea60a4cb0432ce577c31beb009c5c2c49aa2e4eadb217ad8cc09b
[None, None, None, None, None]
>>> [ print(f(bytes(data, encoding='ASCII'))) for data in datas for f in [sha512_224] ]
6ed0dd02806fa89e25de060c19d3ac86cabb87d6a0ddd05c333b84f4
06001bf08dfb17d2b54925116823be230e98b5c6c278303bc4909a8c
3a312b004a593b706790a4a5b25309eb7c83efb85a4d1f0a8440e09e
283bb59af7081ed08197227d8f65b9591ffe1155be43e9550e57f941
37ab331d76f0d36de422bd0edeb22a28accd487b7a8453ae965dd287
[None, None, None, None, None]
>>> [ print(f(bytes(data, encoding='ASCII'))) for data in datas for f in [sha512_256] ]
c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a
3d37fe58435e0d87323dee4a2c1b339ef954de63716ee79f5747f94d974f913f
9423e3863ebb6f22b9464aeb873a39d757ef6b6a87c4bc55642f69052741fc43
10baad1713566ac2333467bddb0597dec9066120dd72ac2dcb8394221dcbe43d
9a59a052930187a97038cae692f30708aa6491923ef5194394dc68d56c74fb21
[None, None, None, None, None]
```

----

- [ ] Because of added new APIs, maybe the version number should be increased to V0.8.0 ?

### CODE REVIEW

**Comment by inkydragon** [comment]
cc @staticfloat , @pixel27

File: src/types.jl [hunk 1]
```diff
@@ -175,6 +195,18 @@ SHA2_384_CTX() = SHA2_384_CTX(copy(SHA2_384_initial_hash_value), 0, zeros(UInt8,
 Construct an empty SHA2_512 context.
 """
 SHA2_512_CTX() = SHA2_512_CTX(copy(SHA2_512_initial_hash_value), 0, zeros(UInt8, blocklen(SHA2_512_CTX)), false)
+"""
+    SHA2_512_224_CTX()
+
+Construct an empty SHA2_512/224 context.
+"""
```
Comment (fingolfin) [inline src/types.jl:202]: Would be nice if these docstrings contained a pointer/link to an authorative place defining the hash (possibly taken from the comments in some of the test files?)

File: src/types.jl [hunk 2]
```diff
@@ -44,6 +44,20 @@ mutable struct SHA2_512_CTX <: SHA2_CTX
     used::Bool
 end
 
+mutable struct SHA2_512_224_CTX <: SHA2_CTX
+    state::Array{UInt64,1}
+    bytecount::UInt128
+    buffer::Array{UInt8,1}
```
Comment (inkydragon) [inline src/types.jl:50]:
```suggestion
    state::Vector{UInt64}
    bytecount::UInt128
    buffer::Vector{UInt8}
```

File: src/types.jl [hunk 3]
```diff
@@ -44,6 +44,20 @@ mutable struct SHA2_512_CTX <: SHA2_CTX
...
+    used::Bool
+end
+
+mutable struct SHA2_512_256_CTX <: SHA2_CTX
+    state::Array{UInt64,1}
+    bytecount::UInt128
+    buffer::Array{UInt8,1}
```
Comment (inkydragon) [inline src/types.jl:57]:
```suggestion
    state::Vector{UInt64}
    bytecount::UInt128
    buffer::Vector{UInt8}
```

**Review by inkydragon — COMMENTED** [review]
Simplify types.

**Review by staticfloat — APPROVED** [review]

File: src/types.jl (hunk 1 above)
Comment (inkydragon) [reply src/types.jl:205]: Done.

**Pull request merged**